pkg/upstream/upstream.go

```diff
@@ -198,3 +220,42 @@ func (up *Upstream) IsNormal() bool {
 func (up *Upstream) IsClosed() bool {
 	return atomic.LoadInt32(&up.status) == closed
 }
+
+func (up *Upstream) unhold() {
+	up.hcMu.mu.Lock()
+	defer up.hcMu.mu.Unlock()
+	up.hcMu.hc--
+	if up.hcMu.hc < 0 {
+		log.Panic("upstream's hc should never less than 0", zap.Uint64("upstreamID", up.ID))
+	}
+	if up.hcMu.hc == 0 && up.hcMu.idleTime.IsZero() {
+		up.hcMu.idleTime = time.Now()
+	}
+}
+
+func (up *Upstream) hold() {
+	up.hcMu.mu.Lock()
+	defer up.hcMu.mu.Unlock()
+	if up.hcMu.hc < 0 {
+		log.Panic("upstream's hc should never less than 0", zap.Uint64("upstreamID", up.ID))
+	}
+	up.hcMu.hc++
+	// reset unusedTime
+	if !up.hcMu.idleTime.IsZero() {
+		up.hcMu.idleTime = time.Time{}
+	}
+}
+
+func (up *Upstream) Release() {
+	up.unhold()
+}
+
+// return true if this upstream idleTime reachs maxIdleDuration.
+func (up *Upstream) shouldClose() bool {
+	up.hcMu.mu.Lock()
+	defer up.hcMu.mu.Unlock()
+	if !up.hcMu.idleTime.IsZero() && time.Since(up.hcMu.idleTime) >= maxIdleDuration {
```

Comment:
Should check if `hc==0` here, rather than idleTime.

Reply:
No, `hc` can be `0` when a upstream is created but is not hold by a `changefeed` or `processor`.